test/extensions/filters/common/ext_authz/ext_authz_http_impl_test.cc

```diff
@@ -480,6 +483,44 @@ TEST_F(ExtAuthzHttpClientTest, CancelledAuthorizationRequest) {
   client_->cancel();
 }
 
+// Test the client when the request times out on an internal timeout.
+TEST_F(ExtAuthzHttpClientTest, AuthorizationInternalRequestTimeout) {
+  initialize("", true);
+  envoy::service::auth::v3::CheckRequest request;
+
+  EXPECT_CALL(async_client_, dispatcher());
+  NiceMock<Event::MockTimer>* timer = new NiceMock<Event::MockTimer>(&async_client_.dispatcher_);
+  EXPECT_CALL(*timer, enableTimer(std::chrono::milliseconds(REQUEST_TIMEOUT), _));
+
+  EXPECT_CALL(async_client_, send_(_, _, _)).WillOnce(Return(&async_request_));
+  client_->check(request_callbacks_, request, parent_span_, stream_info_);
+
+  EXPECT_CALL(async_request_, cancel());
+  EXPECT_CALL(request_callbacks_,
+              onComplete_(WhenDynamicCastTo<ResponsePtr&>(AuthzErrorResponse(CheckStatus::Error))));
+  timer->invokeCallback();
+}
+
+// Test when the client is cancelled with internal timeout.
+TEST_F(ExtAuthzHttpClientTest, AuthorizationInternalRequestTimeoutCancelled) {
+  initialize("", true);
+  envoy::service::auth::v3::CheckRequest request;
+
+  EXPECT_CALL(async_client_, dispatcher());
+  NiceMock<Event::MockTimer>* timer = new NiceMock<Event::MockTimer>(&async_client_.dispatcher_);
+  EXPECT_CALL(*timer, enableTimer(std::chrono::milliseconds(REQUEST_TIMEOUT), _));
+
+  EXPECT_CALL(async_client_, send_(_, _, _)).WillOnce(Return(&async_request_));
+  client_->check(request_callbacks_, request, parent_span_, stream_info_);
+
+  // make sure cancel resets the timer:
+  EXPECT_CALL(async_request_, cancel());
+  bool timer_destroyed = false;
+  timer->timer_destroyed_ = &timer_destroyed;
+  client_->cancel();
+  EXPECT_EQ(timer_destroyed, true);
+}
+
```

Comment:
Should we have a full integration test in test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc etc?